utt/include/globalParams.hpp

```diff
@@ -0,0 +1,50 @@
+// UTT
+//
+// Copyright (c) 2020-2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the LICENSE
+// file.
+
+#pragma once
+#include <string>
+#include <memory>
+namespace libutt {
+class Params;
+}
+namespace libutt::api {
+class GlobalParams {
+  /**
+   * @brief Represents a shared known global UTT params. These parameters includes commitment keys, nullifier parametrs
+   * and more (see uttlib/Params.h)
+   *
+   */
+  static bool initialized;
+
+ public:
+  GlobalParams() = default;
+  /**
+   * @brief Initialize the global parameters object with a default object.
+   *
+   */
+  void init();
+
+  /**
+   * @brief Get the libutt::Params object
+   *
+   * @return const libutt::Params&
+   */
+  const libutt::Params& getParams() const;
+  libutt::Params& getParams();
+  GlobalParams(const GlobalParams& other);
+  GlobalParams& operator=(const GlobalParams& other);
```

Comment:
Couple of things worry me about the code design of this "global params" implementation:
1) You use the non-const version of "getParams()" as a form of a setter.
2) The "init" method uses a static class flag to do the library initialization once - while this is fine, init is also responsible for allocating the memory for params. Later params is assumed that it's always allocated.
3) There's a confusing mix of static and non-static data usage.

What Global params should be to be a "clean" implementation IMO:
1) Global params should be a singleton class, where the instance object is created and initialized only once before first use - this is probably necessary because of how libff/NTL are globally initialized.
2) We might need to customize the creation of global params (by passing extra parameters for libff/NTL or seeds), so we might need an explicit static "create" method called by the user of the library. However besides this, we should still rely on a "instance()" method to get the singleton object.
3) Global params object should have a private constructor so the only way to be created is to use the create/instance public methods.

Reply:
@guyg7 suggested not using a singleton.
However, I reorganized the code such that one needs to run the following steps before starting:
1. initialize the underlying libraries
2. knowing the coin commitment key after generating the secrets for coin signing
3. knowing the registration commitment key after generating the secrets for registration signing
4. create the global params

Reply:
Then why keep the name "Global params" if it's not supposed to be a singleton? Also the initialization step "allocated params once", which made it behave like a singleton, but I see that you changed that.